frontend/src/app/modules/project/components/project-window/project-window.component.ts

```diff
@@ -0,0 +1,284 @@
+                ];
+                break;
+            }
+            case 1:{
+                this.compilerTypes = [
+                    { label: 'V8', value: 2 }
+                ];
+                this.projectTypes = [
+                    { label: 'Console App', value: 0 },
+                ];
+                break;
+            }
+            case 2:{
+                this.compilerTypes = [
+                    { label: 'V8', value: 2 }
+                ];
+                this.projectTypes = [
+                    { label: 'Console App', value: 0 },
+                ];
+                break;
+            }
+            case 3:{
+                this.compilerTypes = [
+                    { label: 'Gc', value: 3 }
+                ];
+                this.projectTypes = [
+                    { label: 'Console App', value: 0 },
+                ];
+                break;
+            }
+        }
+    }
+
+    public getErrorMessage(field: string): string {
+        const control = this.projectForm.get(field);    
+
+        let errorMessage: string;        
+        if (control.hasError('required')) {
+            errorMessage = 'Value is required!';
+        }
+        else if(control.hasError('max')) {
+            errorMessage = `Quantity must be less than ${control.errors.max.max}!`;
+        }
+        else if (control.hasError('minlength')) {
+            errorMessage = `The length must be at least ${control.errors.minlength.requiredLength} letters!`;
+        }
+        else if (control.hasError('maxlength')) {
+            errorMessage = `The length should be no more than ${control.errors.maxlength.requiredLength} letters!`;
+        }
+        else {
+            errorMessage = 'validation error';
+        }
+        return errorMessage;
+    }
+
+    public close() {
+        this.ref.close();
+    }
+        
+    private InitializeProject() {
+        this.projectForm.get('name').setValue(this.projectUpdateStartState.name);
+        this.projectForm.get('description').setValue(this.projectUpdateStartState.description);
+        this.projectForm.get('countOfSavedBuilds').setValue(this.projectUpdateStartState.countOfSaveBuilds);
+        this.projectForm.get('countOfBuildAttempts').setValue(this.projectUpdateStartState.countOfBuildAttempts);
+        this.projectForm.get('color').setValue(this.projectUpdateStartState.color);
+    }
+
+    private SetProjectObjectsFromResponse(resp: HttpResponse<ProjectInfoDTO>): void {
+        const res = resp.body;
+        this.projectUpdateStartState = resp.body;
+        this.projectUpdate.id = res.id;
+        this.projectUpdate.name = res.name;
+        this.projectUpdate.description = res.description;
+        this.projectUpdate.countOfSaveBuilds = res.countOfSaveBuilds;
+        this.projectUpdate.countOfBuildAttempts = res.countOfBuildAttempts;
+        this.projectUpdate.color = res.color;
+        this.projectUpdate.accessModifier = res.accessModifier;
```

Comment:
why do you set every property individually? Better use Object.assign({}, original), or you can use [lodash ](https://lodash.com/docs/4.17.15) clone

Reply:
Because in such a way, i get projectUpdate with more fields, than this class has
Maybe it's because we don't create API method just for getting updateClass and get projectDescription class for it